File: rn-videofeed/ios/VideoFeedPlayerPool.swift
import AVFoundation
import Foundation

/// LRU pool of AVPlayer instances keyed by video id — keeps players warm so
/// scrolling back resumes where the user left off (Reels-style behaviour).
class VideoFeedPlayerPool {
  private struct PooledEntry {
    let player: AVPlayer
    var videoUrl: String
    var lastAccessTime: Date = Date()
  }

  private let maxSize: Int
  private var entries: [String: PooledEntry] = [:]
  private var order: [String] = []

  init(maxSize: Int = 8) {
    self.maxSize = maxSize
  }

  func acquire(videoId: String, videoUrl: String) -> AVPlayer {
    if let entry = entries[videoId], entry.videoUrl == videoUrl {
      touch(videoId)
      print("VideoFeedPlayerPool - Reusing pooled player for \(videoId)")
      return entry.player
    }

    if entries[videoId] != nil {
      removeEntry(videoId)
    }

    while order.count >= maxSize, let oldest = order.first {
      print("VideoFeedPlayerPool - Evicting oldest player: \(oldest)")
      removeEntry(oldest)
    }

    guard let url = URL(string: videoUrl) else {
      fatalError("Invalid video URL: \(videoUrl)")
    }

    let item = AVPlayerItem(url: url)
    let player = AVPlayer(playerItem: item)
    player.isMuted = false

    entries[videoId] = PooledEntry(player: player, videoUrl: videoUrl)
    order.append(videoId)
    print("VideoFeedPlayerPool - Created pooled player for \(videoId) (pool size=\(order.count))")
    return player
  }

  func pause(videoId: String) {
    entries[videoId]?.player.pause()
  }

  func hasPlayer(videoId: String, videoUrl: String) -> Bool {
    guard let entry = entries[videoId] else { return false }
    return entry.videoUrl == videoUrl
  }

  func getPlaybackPosition(videoId: String) -> TimeInterval {
    guard let player = entries[videoId]?.player else { return 0 }
    let seconds = player.currentTime().seconds
    return seconds.isFinite ? seconds : 0
  }

  func releaseAll() {
    print("VideoFeedPlayerPool - Releasing all \(order.count) pooled players")
    for key in order {
      entries[key]?.player.pause()
    }
    entries.removeAll()
    order.removeAll()
  }

  private func touch(_ videoId: String) {
    entries[videoId]?.lastAccessTime = Date()
    order.removeAll { $0 == videoId }
    order.append(videoId)
  }

  private func removeEntry(_ videoId: String) {
    entries.removeValue(forKey: videoId)?.player.pause()
    order.removeAll { $0 == videoId }
  }
}

File: rn-videofeed/android/src/main/java/com/rnvideofeed/VideoFeedPlayerPool.kt
package com.rnvideofeed

import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.media3.common.MediaItem
import androidx.media3.common.MimeTypes
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer

/**
 * Keeps a small LRU pool of ExoPlayer instances keyed by video id so scrolling
 * back resumes playback where the user left off (Reels-style behaviour).
 */
class VideoFeedPlayerPool(
    private val context: Context,
    private val maxSize: Int = 8
) {
    private val tag = "VideoFeedPlayerPool"

    private data class PooledEntry(
        val player: ExoPlayer,
        var videoUrl: String,
        var lastAccessTime: Long = System.currentTimeMillis()
    )

    private val entries = LinkedHashMap<String, PooledEntry>(maxSize, 0.75f, true)

    fun acquire(videoId: String, videoUrl: String): ExoPlayer {
        entries[videoId]?.let { entry ->
            if (entry.videoUrl == videoUrl) {
                entry.lastAccessTime = System.currentTimeMillis()
                Log.d(tag, "Reusing pooled player for $videoId at ${entry.player.currentPosition}ms")
                return entry.player
            }
            removeEntry(videoId)
        }

        while (entries.size >= maxSize) {
            val oldestKey = entries.keys.first()
            Log.d(tag, "Evicting oldest player: $oldestKey")
            removeEntry(oldestKey)
        }

        val player = ExoPlayer.Builder(context).build().apply {
            repeatMode = Player.REPEAT_MODE_ONE
            volume = 1.0f
        }
        val uri = Uri.parse(videoUrl)
        val mimeType = when {
            videoUrl.endsWith(".m3u8", ignoreCase = true) -> MimeTypes.APPLICATION_M3U8
            videoUrl.endsWith(".mp4", ignoreCase = true) -> MimeTypes.VIDEO_MP4
            else -> null
        }
        val mediaItem = MediaItem.Builder()
            .setUri(uri)
            .apply { if (mimeType != null) setMimeType(mimeType) }
            .build()
        player.setMediaItem(mediaItem)
        player.prepare()

        entries[videoId] = PooledEntry(player, videoUrl)
        Log.d(tag, "Created new pooled player for $videoId (pool size=${entries.size})")
        return player
    }

    fun pause(videoId: String) {
        entries[videoId]?.player?.pause()
    }

    fun hasPlayer(videoId: String, videoUrl: String): Boolean {
        val entry = entries[videoId] ?: return false
        return entry.videoUrl == videoUrl
    }

    fun getPlaybackPosition(videoId: String): Long {
        return entries[videoId]?.player?.currentPosition ?: 0L
    }

    fun releaseAll() {
        Log.d(tag, "Releasing all ${entries.size} pooled players")
        entries.values.forEach { it.player.release() }
        entries.clear()
    }

    private fun removeEntry(videoId: String) {
        entries.remove(videoId)?.player?.release()
    }
}
